### PR TITLE
Refactor and speedup url-rewrite

### PR DESCRIPTION
**Pull request by buger**
- Added new `replaceTykVariables` function, which replaces all
`$tyk_context` and `$tyk_meta` variables in string`
- URLRewrite regexp now cached. Previously it was compiling on each
call.

### CODE REVIEW

File: mw_url_rewrite.go [hunk 1]
```diff
@@ -17,17 +17,14 @@ import (
...
-		return "", err
+	if meta.MatchRegexp == nil {
+		return "", fmt.Errorf("URLRewrite regexp error", meta.MatchPattern)
 	}
 	path := r.URL.String()
 	log.Debug("Inbound path: ", path)
 	newpath := path
 
-	result_slice := mp.FindAllStringSubmatch(path, -1)
+	result_slice := meta.MatchRegexp.FindAllStringSubmatch(path, -1)
```
Comment (dencoded) [inline mw_url_rewrite.go]: `snake_case` for var names looks weird in Go

File: mw_url_rewrite.go [hunk 2]
```diff
@@ -157,51 +153,64 @@ func urlRewrite(meta *apidef.URLRewriteMeta, r *http.Request) (string, error) {
...
+
+func replaceTykVariables(r *http.Request, in string, escape bool) string {
 	contextData := ctxGetData(r)
 
-	dollarMatch := regexp.MustCompile(`\$tyk_context.([A-Za-z0-9_\-\.]+)`)
-	replace_slice := dollarMatch.FindAllStringSubmatch(rewriteToPath, -1)
-	for _, v := range replace_slice {
-		contextKey := strings.Replace(v[0], "$tyk_context.", "", 1)
+	if strings.Contains(in, contextLabel) {
+		replace_slice := dollarMatch.FindAllStringSubmatch(in, -1)
```
Comment (dencoded) [inline mw_url_rewrite.go]: one more `snake_case` spotted

File: mw_url_rewrite.go [hunk 3]
```diff
@@ -157,51 +153,64 @@ func urlRewrite(meta *apidef.URLRewriteMeta, r *http.Request) (string, error) {
...
-	if session := ctxGetSession(r); session != nil {
+	session := ctxGetSession(r)
+	if session == nil {
+		return in
+	}
 
-		metaDollarMatch := regexp.MustCompile(`\$tyk_meta.(\w+)`)
-		metaReplace_slice := metaDollarMatch.FindAllStringSubmatch(rewriteToPath, -1)
+	if strings.Contains(in, metaLabel) {
+		metaReplace_slice := metaDollarMatch.FindAllStringSubmatch(in, -1)
```
Comment (dencoded) [inline mw_url_rewrite.go]: and one more snake

File: mw_url_rewrite.go [hunk 4]
```diff
@@ -157,51 +153,64 @@ func urlRewrite(meta *apidef.URLRewriteMeta, r *http.Request) (string, error) {
...
+var metaDollarMatch = regexp.MustCompile(`\$tyk_meta.([A-Za-z0-9_\-\.]+)`)
+
+func replaceTykVariables(r *http.Request, in string, escape bool) string {
 	contextData := ctxGetData(r)
 
-	dollarMatch := regexp.MustCompile(`\$tyk_context.([A-Za-z0-9_\-\.]+)`)
-	replace_slice := dollarMatch.FindAllStringSubmatch(rewriteToPath, -1)
-	for _, v := range replace_slice {
-		contextKey := strings.Replace(v[0], "$tyk_context.", "", 1)
+	if strings.Contains(in, contextLabel) {
```
Comment (dencoded) [inline mw_url_rewrite.go]: we could eliminate a couple of extra ops if move `contextData := ctxGetData(r)` inside this if

File: mw_url_rewrite.go [hunk 5]
```diff
@@ -157,51 +153,64 @@ func urlRewrite(meta *apidef.URLRewriteMeta, r *http.Request) (string, error) {
...
 	// Meta data from the token
-	if session := ctxGetSession(r); session != nil {
+	session := ctxGetSession(r)
+	if session == nil {
+		return in
+	}
 
-		metaDollarMatch := regexp.MustCompile(`\$tyk_meta.(\w+)`)
-		metaReplace_slice := metaDollarMatch.FindAllStringSubmatch(rewriteToPath, -1)
+	if strings.Contains(in, metaLabel) {
```
Comment (dencoded) [inline mw_url_rewrite.go]: also could get rid of a couple of ops if we move `if session := ctxGetSession(r); ...` inside this if

File: mw_modify_headers.go [hunk 6]
```diff
@@ -87,7 +38,9 @@ func (t *TransformHeaders) ProcessRequest(w http.ResponseWriter, r *http.Request
 
 	// Add
 	if len(vInfo.GlobalHeaders) > 0 {
```
Comment (dencoded) [inline mw_modify_headers.go]: seems like extra check as iteration over empty map will be no operation as well

File: mw_url_rewrite.go [hunk 7]
```diff
@@ -17,17 +17,14 @@ import (
 )
 
 func urlRewrite(meta *apidef.URLRewriteMeta, r *http.Request) (string, error) {
-	// Find all the matching groups:
-	mp, err := regexp.Compile(meta.MatchPattern)
-	if err != nil {
-		log.Debug("Compilation error: ", err)
-		return "", err
+	if meta.MatchRegexp == nil {
+		return "", fmt.Errorf("URLRewrite regexp error", meta.MatchPattern)
```
Comment (dencoded) [inline mw_url_rewrite.go]: we need to add placeholder into format string - this one was caught by my IDE :)

**Comment by buger** [comment]
@dencoded existing benchmark won't show anything because it does not use URL rewrite plugins. Smth we need to address in future.

**Pull request merged**